tests/ChatTests/Local.hs

```diff
@@ -0,0 +1,128 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE PostfixOperators #-}
+
+module ChatTests.Local where
+
+import ChatClient
+import ChatTests.Utils
+import System.Directory (copyFile, doesFileExist)
+import Test.Hspec
+import System.FilePath ((</>))
+import Data.Time (getCurrentTime)
+import ChatTests.ChatList (getChats_)
+import Data.Time.Format.ISO8601 (iso8601Show)
+
+chatLocalTests :: SpecWith FilePath
+chatLocalTests = do
+  describe "note folders" $ do
+    it "create folders, add notes, read, search" testNotes
+    it "switch users" testUserNotes
+    it "preview pagination for notes" testPreviewsPagination
+    it "stores files" testFiles
+
+testNotes :: FilePath -> IO ()
+testNotes tmp = withNewTestChat tmp "alice" aliceProfile $ \alice -> do
+  alice ##> "/contacts"
+  -- not a contact
+
+  alice /- "keep in mind"
+  alice ##> "/tail"
+  alice <# "- keep in mind"
+  alice ##> "/chats"
+  alice <# "- keep in mind"
+  alice ##> "/? keep"
+  alice <# "- keep in mind"
+
+  alice #$> ("/_read chat $1 from=1 to=100", id, "ok")
+  alice ##> "/_unread chat $1 on"
+  alice <## "ok"
+
+  alice ##> "/_delete item $1 1 internal"
+  alice <## "message deleted"
+  alice ##> "/tail"
+  alice ##> "/chats"
+
+  alice /- "ahoy!"
+  alice ##> "/_update item $1 1 text Greetings."
+  alice ##> "/tail -"
+  alice <# "- Greetings."
+
+testUserNotes :: FilePath -> IO ()
+testUserNotes tmp = withNewTestChat tmp "alice" aliceProfile $ \alice -> do
+  alice /- "keep in mind"
+  alice ##> "/tail"
+  alice <# "- keep in mind"
+
+  alice ##> "/create user secret"
+  alice <## "user profile: secret"
+  alice <## "use /p <display name> to change it"
+  alice <## "(the updated profile will be sent to all your contacts)"
+
+  alice ##> "/tail"
+
+  alice ##> "/_delete item $1 1 internal"
+  alice <## "chat db error: SENoteFolderNotFound {noteFolderId = 1}"
+
+testPreviewsPagination :: FilePath -> IO ()
+testPreviewsPagination tmp = withNewTestChat tmp "alice" aliceProfile $ \alice -> do
+  tsS <- iso8601Show <$> getCurrentTime
+  alice /- "first"
+  tsM <- iso8601Show <$> getCurrentTime
+  alice /- "last"
+  tsE <- iso8601Show <$> getCurrentTime
+
+  -- there's only one folder that got updated after tsM and before tsE
+  getChats_ alice "count=3" [("-", "last")]
+  getChats_ alice ("after=" <> tsE <> " count=10") []
+  getChats_ alice ("after=" <> tsS <> " count=10") [("-", "last")]
+  getChats_ alice ("before=" <> tsM <> " count=10") []
+  getChats_ alice ("before=" <> tsE <> " count=10") [("-", "last")]
+  getChats_ alice ("before=" <> tsS <> " count=10") []
```

Comment:
this only tests chat list pagination, should also test chat pagination (pagination of chat items inside chat). also when you always take count=10 it doesn't show that "after" behaves correctly and not the same as "last" - chat should have more items and test should check after some early item but not until the end

Reply:
There's only one "chat" here, so any `count` above `1` is the same.

Reply:
exactly, that's why I'm saying that chat should have more items - currently test doesn't check different cases of chat pagination

Reply:
also as far as I could tell there're only tests for chat list pagination and no tests of chat pagination

Reply:
I added chat items test.

This one I don't get. This test has only notes as mixing with the others is tested in the combined chat list test.
There are 2 notes and some of the `[]` results are where the notes folder got hidden entirely instead of showing "first" note. And this is correct behavior since the folder timestamp is indeed outside the range despite having older items in there (otherwise chat list pagination wouldn't work, showing all kinds of stuff all the time).

Reply:
> I added chat items test.

ok, I was referring to the initial comment before this test was added